[authn] use authlib client+interceptors for in-proc mode

drclau opened the pull request:
Part of https://github.com/grafana/identity-access-team/issues/589

**What is this feature?**

Use `authlib`'s gRPC interceptors and authenticators for in-proc mode.

![image](https://github.com/user-attachments/assets/1712cc73-2764-4b62-883f-3767797b9680)


Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

drclau committed:
Add authlib gRPC authenticators for in-proc mode

ryantxu commented:
With this are we also able to remove: https://github.com/grafana/grafana/tree/main/pkg/storage/unified/resource/grpc

ryantxu commented:
This PR currently fails when the aggregator is enabled:
```
[feature_toggles]
kubernetesAggregator = true
```

drclau committed:
Fixed apistore tests

drclau committed:
aesthetics

drclau commented:
> For in process, I thought we could get away with a noop Authenticator where we just use the same context -- but it looks like the inproc grpc wrapper actually requires we use a new context 😢
> 
> I eventually landed on #93160 as an alternative that skips gRPC all together

@ryantxu, your variant seems more efficient. I propose we prioritize getting your PR merged, and meanwhile we keep this one as a backup (in draft mode; we can close it later).
And on our side we'll focus on gRPC modes, in a separate PR.

drclau committed:
cleanup

drclau commented:
> This PR currently fails when the aggregator is enabled:
> 
> ```
> [feature_toggles]
> kubernetesAggregator = true
> ```

@ryantxu, I've done some digging into what's going on when `kubernetesAggregator = true`, and traced it back to [this code](https://github.com/grafana/grafana/blob/21d4a4f49ee4d74119089c5c2e2d2d6878f4faa0/pkg/apiserver/endpoints/filters/requester.go#L37-L71).
TL;DR, a `StaticRequester` with type `claims.TypeServiceAccount` is instantiated (plus some other important fields are set), and is injected into `context`. This `context` eventually reaches the gRPC client interceptor that is called whenever `ResourceStoreClient` is used.

This worked with the old implementation, because the fields were taken one by one and injected into the gRPC metadata on client side. The reverse was done on server side, and the `StaticRequester` was reconstructed. With the new implementation, we rely solely on the ID token to extract the identity, which has the advantage of sharing most of the implementation and flows with the remote gRPC implementation.

Probably the most elegant solution to making `kubernetesAggregator` work with the implementation in this PR is to sign the `StaticRequester` and thus get an ID token for it, too. It's also very simple, I just need to figure out how to make `IDService` available in this part of the code (any suggestions are welcome!).

WDYT?

drclau committed:
Fix for `kubernetesAggregator` mode (in progress)

- [x] implement `StaticRequester` signing in `apiserver` in-proc
- [ ] handle standalone `apiserver` mode
- [ ] update `grafana-enterprise`

drclau committed:
Merge branch 'main' into drclau/authn-inproc-interceptors

drclau committed:
update NewLocalResourceClient() to return ResourceClient instead of ResourceStoreClient

ryantxu commented on pkg/apiserver/endpoints/filters/requester.go:
```diff
@@ -64,6 +68,11 @@ func WithRequester(handler http.Handler) http.Handler {
 				}
 			}
 
+			requester, err = signStaticRequester(ctx, idService, requester.(*identity.StaticRequester))
```
🤔  maybe we can add the signing to the grpc interceptor 🤔

drclau replied:
✅

drclau committed:
revert signing of StaticRequester in filters.WithRequester()

drclau committed:
implement StaticRequester signing in the unified resource client

- [x] when the `claims.AuthInfo` value type is `identity.StaticRequester`, and there's no ID token set, create an internal token and sign it with symmetrical key. This is a workaround for go-jose not offering the possibility to create an unsigned token.
- [x] update IDClaimsWrapper to support the scenario above
- [x] Switch to using `claims.From()` in `dashboardSqlAccess.SaveDashboard()`

drclau committed:
Merge branch 'main' into drclau/authn-inproc-interceptors

drclau committed:
cleanup

gamab reviewed (approved):
LGTM from a code perspective! Good job following up on this with consistency. 

It's a shame we had to deviate a bit more than expected from the cloud flow here. Hopefully on-prem service tokens will save the day... one day :)

gamab commented on pkg/storage/unified/apistore/store_test.go:
```diff
@@ -32,13 +32,41 @@ func init() {
+		//   "type": "user",
+		//   "username": "testuser"
+		// }
+		idToken := "eyJhbGciOiJFUzI1NiIsInR5cCI6Imp3dCJ9." +
```
Is this still needed in the end?

gamab committed:
Remove test token

drclau committed:
cleanup: revert store_test.go to main version

drclau committed:
tests: create ID tokens for user.SignedInUser

ryantxu reviewed (approved):
nice
